content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md

```diff
@@ -0,0 +1,864 @@
+
+```bash
+Hello, World!
+```
+
+Once the testing is complete, you can destroy the cluster using this command.
+
+```bash
+kind delete cluster --name wasm-demo
+```
+##### KWasm Operator
+
+_[Kwasm](https://kwasm.sh/)_ is a Kubernetes Operator that adds WebAssembly support to your Kubernetes nodes. The operator uses the _[kwasm-node-installer](https://github.com/KWasm/kwasm-node-installer)_ project to modify the underlying Kubernetes nodes.
+
+In this section, I will demostrate how to use Kwasm Operator to add Wasm support to Kubernetes nodes automatically. To enable Wasm support on a particular node, simply add the annotation `kwasm.sh/kwasm-node=true` on that node. This will trigger Kwasm to create a Job to deploy the necessary binary files needed to run Wasm on the node. Additionally, containerd's configuration will be modified accordingly.
+
+{{< figure src="05-kwasm-operator.svg" alt="The diagram illustrates that when Kwasm Operator detects the annotation kwasm.sh/kwasm-node=true on a Kubernetes node, it will create a Job to enable Wasm support on that node." caption="Figure 5: Kwasm Operator" >}}
+
+Create a single-node Kubernetes cluster using kind.
+
+```bash
+kind create cluster --name kwasm-demo
+```
+
+A Helm chart is available to easily install the Kwasm operator. Install the Kwasm Operator using helm and enable Wasm support for the node `kwasm-demo-control-plane` by adding the annotation `kwasm.sh/kwasm-node=true`.
+
+```bash
+# Add HELM repository if not already done
+helm repo add kwasm http://kwasm.sh/kwasm-operator/
+# Install KWasm operator
+helm install -n kwasm --create-namespace kwasm-operator kwasm/kwasm-operator
+# Provision Nodes
+kubectl annotate node kwasm-demo-control-plane kwasm.sh/kwasm-node=true
+```
+
+Add label `runtime=crun` on the node.
+
+```bash
+kubectl label nodes kwasm-demo-control-plane runtime=crun
+```
+
+Create a RuntimeClass resource named `crun` to use the crun handler automatically set up by Kwasm in containerd, the `scheduling.nodeSelector` property sends pod to nodes with the `runtime=crun` label.
+
+```yaml
+apiVersion: node.k8s.io/v1
+kind: RuntimeClass
+metadata:
+  name: crun
+scheduling:
+  nodeSelector:
+    runtime: crun
+handler: crun
+```
+
+Set `.spec.runtimeClassName` for the pod to target the pod at the `crun` RuntimeClass. This will ensure the pod gets assigned to a node and runtime specified in the `crun` RuntimeClass. Additionally, set the annotation `module.wasm.image/variant: compat` to inform crun that this is a Wasm workload.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: wasm-demo-app
+  annotations:
+    module.wasm.image/variant: compat
+spec:
+  runtimeClassName: crun
+  containers:
+  - name: wasm-demo-app
+    image: docker.io/cr7258/wasm-demo-app:v1
+```
+
+You can use `port-forward` to forward traffics to your local machine for access.
+
+```bash
+kubectl port-forward pod/wasm-demo-app 8080:8080
+```
+
+Open a new terminal, send a request to the service.
```

Comment:
```suggestion
Send a request to the service running in the test Pod inside Kubernetes:
```

Reply:
I have made the service available locally using `kubectl port-forward`, and I can access it through the URL http://localhost:8080. So there is no need to send the request in the test Pod inside Kubernetes?